### ISSUE TITLE
[semi] incorrectly requires semicolon after `export default interface`

### ISSUE THREAD

**Issue by MayhemYDG** [typescript-eslint/typescript-eslint#123]
**What code were you trying to parse?**

```ts
export default interface test {}
```

**What did you expect to happen?**
`semi` should ignore the interface block.
**What actually happened?**
`semi` wants a semicolon after the block, which then triggers the `no-extra-semi` rule.

It works fine without `default`:
```ts
export interface test1 {}
interface test2 {}
```
**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.0.0` |
| `TypeScript`                | `3.2.4` |
| `ESLint`                    | `5.12.1` |
| `node`                      | `11.7.0` |
| `npm`                       | `6.5.0` |

**Comment by bradzacher** [comment]
`TSInterfaceDeclaration` by itself doesn't trigger a semi check purely because it doesn't match the base rule's matchers.

https://github.com/eslint/eslint/blob/faf3c4eda0d27323630d0bc103a99dd0ecffe842/lib/rules/semi.js#L320-L324

The base rule has a specific case for `ExportDefaultDeclaration`, and only skips the semi check if it's a class or function declaration.

We might have to override the rule for this one.

Temp workaround is to define your interface separately:
```ts
interface Foo {}
export default Foo;
```

**Comment by MayhemYDG** [comment]
@bradzacher could this possibly be fixed in eslint by making the `ExportDefaultDeclaration` check use a whitelist instead of a blacklist?

**Comment by bradzacher** [comment]
Probably would fix it, but the problem is:
- there is a large number of nodes that would need to be whitelisted
- it would cause false positives as new language features get supported by the base ESTree representation

We can bring this rule internally and extend its so that it works for TS nodes.

**Comment by robertmain** [comment]
I don't suppose this is related to https://github.com/eslint/eslint/issues/9399 is it?

I'm seeing it in my app too.


```typescript
export default interface AudioFactory{
    getSong(id: string): Promise<Song>;
};
```

I get `missing semicolon`

**Comment by JamesHenry** [comment]
Fixed by #461

**Comment by mytee306** [comment]
> Fixed by #461

I've tried:
- `"@typescript-eslint/semi": "error"`
- `"@typescript-eslint/semi": "[2, "always"]"`

Unfortunately, none of which work. Is something missing?

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/releases

Hasn't been released to master yet.
Try the canary tag on npm for builds directly from master if you don't want to wait for a main release.

**Comment by bjerkins** [comment]
@bradzacher tried `1.7.1-alpha.16`, marked as canary, but I still have the same problem 
![image](https://user-images.githubusercontent.com/2454547/56949179-7d864880-6b32-11e9-8ff1-f1c094c87a96.png)

**Comment by bradzacher** [comment]
@bjerkins  - you still have the base semi rule turned on, which is what causes the first error.
you cannot have both the base rule and our rule turned on, as the base rule will false positive.
see the docs on semi: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/semi.md

the second one might be a bug, but it's hard to tell because you've clipped the code out.

**Comment by eyedean** [comment]
I had the same problem as @bjerkins  and just confirmed adding the following from @bradzacher's link fixed it.  Thanks, Brad!
```json
  "semi": "off",
  "@typescript-eslint/semi": ["error"]
```

**Comment by audacioustux** [comment]
any update on this?  
```
export default interface RConfig {
    a: string;
};;; <-- adds semicolon on --fix everytime

```

config: 
```
"semi: ["warn", "always"],
"@typescript-eslint/semi": ["warn", "always"],
"no-extra-semi": 0,
"semi-style": ["error", "last"],
"semi-spacing": ["error", { "before": false, "after": true }],
```

**Comment by bradzacher** [comment]
why have you got both?
```
"semi: ["warn", "always"],
"@typescript-eslint/semi": ["warn", "always"],
```

This will ofc cause a problem because both rules are going to try and add a semi.

Please read the rule doc, which tells you to explicitly NOT do this.
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/semi.md

**Comment by audacioustux** [comment]
> why have you got both?
> 
> ```
> "semi: ["warn", "always"],
> "@typescript-eslint/semi": ["warn", "always"],
> ```
> 
> This will ofc cause a problem because both rules are going to try and add a semi.
> 
> Please read the rule doc, which tells you to explicitly NOT do this.
> https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/semi.md

sorry for deleting my last comment... just figured out what I'm messing with and didn't see someone already seen and replied lol ... but `export default interface RConfig {}` not giving me any error for not putting a semicolon at the end now... any solution for this?
